### ISSUE TITLE
How is the final candle in backtest data.history calculated

### ISSUE THREAD

**Issue by gatapia** [enigmampc/catalyst#266]
I am very confused about the handling of the final candle from `data.history` during backtests.  Let me try to explain with code that reproduces the issue:
```
import pytz
from datetime import datetime, timedelta
from catalyst.api import symbol
from catalyst.utils.run_algo import run_algorithm
from utils.data import load_coin_data_from_csv

exchange_name='bitfinex'
coin = 'btc'
base_currency='usd'
n_candles = 5
resample_frequency_minutes = 5

def initialize(context):        
    context.symbol = symbol('%s_%s' % (coin, base_currency))    
    # very simple method that reads data from generated csv file
    context.all_data = load_coin_data_from_csv('btc', exchange_name, 'minute', from_date=datetime(2017, 6, 1, 0, 0, 0, 0, pytz.utc), resample='%dT' % resample_frequency_minutes)

def handle_data(context, data):    
    bundle_data = context.all_data[data.current_dt - timedelta(minutes=(n_candles - 1)*resample_frequency_minutes):][:n_candles]['close']
    backtest_data = data.history(context.symbol, 'close', bar_count=n_candles, frequency='%dT' % resample_frequency_minutes)
    print('current_time:', data.current_dt)
    
    print('bundle_data:')
    print(bundle_data)
    
    print('backtest_data:')
    print(backtest_data)
    raise Exception('fail')

run_algorithm(
            initialize=initialize,
            handle_data=handle_data,
            exchange_name=exchange_name,
            base_currency=base_currency,
            algo_namespace='investigate_mismatching_final_candle',
            live=False,
            data_frequency='minute',
            capital_base=3000,
            simulate_orders=True,
            start = datetime(2018, 2, 2, 0, 0, 0, 0, pytz.utc),
            end = datetime(2018, 2, 20, 0, 0, 0, 0, pytz.utc))
```
Now for btc_usd on bitfinex I get the following 2 datasets:
```
current_time: 2018-02-02 00:00:00+00:00

bundle_data:
2018-02-01 23:40:00    9306.2
2018-02-01 23:45:00    9270.0
2018-02-01 23:50:00    9216.9
2018-02-01 23:55:00    9181.0
2018-02-02 00:00:00    9150.0 # not matching
Freq: 5T, Name: close, dtype: float64

backtest_data:
2018-02-01 23:35:00+00:00    9269.0
2018-02-01 23:40:00+00:00    9306.2
2018-02-01 23:45:00+00:00    9270.0
2018-02-01 23:50:00+00:00    9216.9
2018-02-01 23:55:00+00:00    9181.0
2018-02-02 00:00:00+00:00    9182.1# not matching
Freq: 5T, Name: TradingPair(98871 [btc_usd]), dtype: float64
```

For btc_usdt on poloniex I get:
```
current_time: 2018-02-02 00:00:00+00:00

bundle_data:
2018-02-01 23:40:00    9297.003598
2018-02-01 23:45:00    9290.418766
2018-02-01 23:50:00    9250.762957
2018-02-01 23:55:00    9199.570000
2018-02-02 00:00:00    9168.790860 # not matching
Freq: 5T, Name: close, dtype: float64

backtest_data:
2018-02-01 23:35:00+00:00    9279.622798
2018-02-01 23:40:00+00:00    9297.003598
2018-02-01 23:45:00+00:00    9290.418766
2018-02-01 23:50:00+00:00    9250.762957
2018-02-01 23:55:00+00:00    9199.570000
2018-02-02 00:00:00+00:00    9199.570000 # not matching
Freq: 5T, Name: TradingPair(374465 [btc_usdt]), dtype: float64
```

Firstly `data.history` returns 6 candles from a request for 5.  This has already been reported I think.

But more importantly: You can see the final candle of both exchanges do not match.  I would expect the final close price @ time T to be the same close price as the bundle data at time T.

@fredfortier explained a few issues relating to the final candle in bug #230 but I don't think that applies here.  Can someone please explain to me what is happening? Is this a bug?  Are you returning a previous close price assuming that time T is slightly in the future?  Is this zipline slippage/black magic?

My ultimate objective is to pre-load all my model predictions at the beginning of my backtest for performance. But I cannot do that if the data from bundles does not match what is returned by `data.history`.  

Cheers.

**Comment by fredfortier** [comment]
Can you please output your catalyst version number? I just want to make
sure that we are looking at an output from 0.5.3.

**Comment by gatapia** [comment]
Doing a 
```
import catalyst
print(catalyst.__version__)
```
prints: `0.5.3+84.gcb4668f` which I think is the current development branch as of right now.

**Comment by fredfortier** [comment]
Just eyeballing the results, I don't think that there is an issue. You are resampling your `backtest_data` in advance in the initialize method. The label of each candle is left-bound. This means that the last candle goes between 0:00 and 0:04:59. Since you are requesting candles at 0:00, I'd expect the close of 0:00 to be the same as the close of 23:55. I also expect your pre-computed `backtest_data` to contain data in the future since your are including the 0:00-0:04:59 candle in your range. If `bundle_data` resulted in the same value, it would include data in the future. Let me know if you think that I'm missing something. In this instance, the closest candle to current is the candle immediately before the current time, not the one corresponding to the current time.

**Comment by fredfortier** [comment]
@lenak25 can you please follow-up on this issue based on our discussion?

**Comment by gatapia** [comment]
Let me see if I understand. At: 2018-02-02 00:00 I request 5 bars from `data.history`, the final 3 bars are:
```
2018-02-01 23:50:00+00:00    9216.9
2018-02-01 23:55:00+00:00    9181.0
2018-02-02 00:00:00+00:00    9182.1
```

The second bar (23:55) shows the close price at 23:59 and the last bar (00:00) shows the close price at (00:00).  This is problematic for algos since all bars show movement over 5 minutes but the last bar (with the most important value as it has the most recent price) only shows movement for an arbitrary length of time.

Wouldn’t the following be more accurate?

Let’s assume we request 2 bars at 00:00.  Then obviously, the last bar should show 9182.1 since that is the most accurate, most up to date price at 00:00, the first bar should then be the close price at 23:55 (9204).  

I.e. The resample label and close should be on the right rather than the left.  This would mean no more partial candles and values always in correct intervals.

Note: This assumes that the exchange APIs cooperate and give you the same data when in live mode.  As the most important thing is to ensure backtest is representative of live trading.  

Reference (minute data from bundle):
```
2018-02-01 23:55:00    9204.00000
2018-02-01 23:56:00    9165.72858
2018-02-01 23:57:00    9194.00000
2018-02-01 23:58:00    9210.00000
2018-02-01 23:59:00    9181.00000
2018-02-02 00:00:00    9182.10000
2018-02-02 00:01:00    9184.00000
2018-02-02 00:02:00    9186.10000
2018-02-02 00:03:00    9189.00000
2018-02-02 00:04:00    9150.00000
2018-02-02 00:05:00    9110.00000
```

**Comment by fredfortier** [comment]
We could label on the right but, as your pointed out, we'd have a consistency issue with exchanges which label on the left. We would still have a partial candle because we would still resample on the clock. For example, if it's 00:02, the last candle would still be 2 minutes worth of data. Again, this how exchanges build their candles.

We also considered removing the last candle. In backtest mode with 1T frequency, this would seems like the obvious thing do. If the clock says 00:00:15, we can simply make the last candle 23:59 without the risk of missing trades between 23:59:59 and 00:00:15. Not so in live mode however. While missing a few seconds worth of trades may be acceptable, what if it's 00:02 and your sampling timeframe is 5T? Removing the partial candle would cut 2 minutes worth of trades both in backtest and live.

**Comment by gatapia** [comment]
Understood, its a difficult one.  I agree partial candle is required to simulate live exchanges if this is the case.

I will use 1m candles resampled to 5M (labels on right) on bitfinex and use 5m candles (throwing out last partial) on polo (since they don't do 1M candles). 

Messy

**Comment by gatapia** [comment]
Sorry, I think I found another issue with the final partial candle, see this:
```
def handle_data_polo_partial_candles(context, data):        
    # all I do is print the current last 2 candles (5T)
    print('\nnow: %s\n%s' % (data.current_dt, data.history(symbol('btc_usdt'), ['close'], bar_count=2, frequency='5T')))
    if not hasattr(context, 'i'): context.i = 0
    context.i += 1
    if context.i > 3: raise Exception('stop')
        
run_algorithm(initialize=lambda ctx: True, handle_data=handle_data_polo_partial_candles,
            exchange_name='poloniex', base_currency='usdt', algo_namespace='ns',
            live=False, data_frequency='minute', capital_base=3000, start = datetime(2018, 2, 2, 0, 0, 0, 0, pytz.utc),
            end = datetime(2018, 2, 20, 0, 0, 0, 0, pytz.utc))
```
This gives the following output:
```
now: 2018-02-02 00:00:00+00:00
                                 close
2018-02-01 23:50:00+00:00  9250.762957
2018-02-01 23:55:00+00:00  9199.570000
2018-02-02 00:00:00+00:00  9199.570000

now: 2018-02-02 00:01:00+00:00
                               close
2018-02-01 23:55:00+00:00  9199.5700
2018-02-02 00:00:00+00:00  9199.5736

now: 2018-02-02 00:02:00+00:00
                                 close
2018-02-01 23:55:00+00:00  9199.570000
2018-02-02 00:00:00+00:00  9179.934131

now: 2018-02-02 00:03:00+00:00
                             close
2018-02-01 23:55:00+00:00  9199.57
2018-02-02 00:00:00+00:00  9200.00
```

So we see at 00:00:00 we have an additional partial candle which is inline with what @fredfortier explained above, i.e. the 2 requested candles +  1 partial. However, the candles at 00:01:00 - 00:03:00 do not have 2 requested candles.  They have 1 + 1 partial, you can tell its partial because it changes on every frame.  

Note: I am using develop branch updated 1 hour ago.

**Comment by fredfortier** [comment]
Looking into this now.

**Comment by lenak25** [comment]
Hi @gatapia , I see what you talking about and I think I understand the root cause. 
On a fixed interval (like 00:00) you get one candle more, meaning 3 candles (2 full + 1 partial) while for other cases you get just 2 candles (1 full + 1 partial).
It is very similar to issue #260 but not exactly the same, as that related to live fetching and this one relates to backtesting.
Just to be sure we are all on the same page: what do you expect to see in the case above: always get 2 candles (1 full + 1 partial) or to always get 2 candles (2 full + 1 partial).
My reasoning is to return always 2 bars as requested - meaning always 1 full + 1 partial - but I would like to be sure we all agree on that.

**Comment by gatapia** [comment]
I don't mind, as long as its consistent.  I would assume 2 full + 1 partial (partial only if applicable?).  Again, I don't think it matters as long as its consistent.

**Comment by zackgow** [comment]
This has been a confusing issue for me as well. Right-labeling seems much more intuitive to me, but it makes sense to keep them left-labeled for consistency.

It makes the most sense that requesting 2 candles should yield 2 candles (1 full + 1 partial), otherwise one would have to write custom logic to choose the appropriate index based on whether or not the data was requested on some fixed interval.

**Comment by lenak25** [comment]
Thanks @gatapia and @zackgow. 
We will fix it to always return 2 candles (1 full + 1 partial) for also being consistent with the live candles fetch.

**Comment by aetherqi** [comment]
Shouldn't calls to `data.history` return only complete historical candles? with partial candles returned by `data.current`?

**Comment by aetherqi** [comment]
So am I right in saying that prior to 0.5.3 the last bar returned by `data.history` (minute data with frequency > 1) was actually the future closing price for that bar? No wonder my algo was performing so well on 0.5.2.

**Comment by zackgow** [comment]
@aetherqi Yes, there was look-ahead bias prior to 0.5.3.

**Comment by xbionicx** [comment]
I agree with @aetherqi , shouldn't `data.history` returns complete candles? It would be more consistent

**Comment by fredfortier** [comment]
We agree. We're trying to align with Zipline when at all possible, but we'll deviate a bit on this one. In Zipline, the last candle of each daily frame end at 23:59 of the previous day. This works great because the current candle is always the last complete one. For the minute frame on the other hand, the last candle is matches the current minute (e.g. 0:00, 0:01, 0:02). We are changing this to 23:59:59, 0:0:59, 0:01:59. And yes, we agree that `data.current()` should suffice to compensate for the lack of a partial candle in live mode.

We are making this change on the "new_exchange_config" branch where in which we are also adding some new data bundles. We've been delayed releasing this because it required more data curation than expected, but we are going to release any day now. We'll do another release in between (probably today). This one will follow shortly.

**Comment by gatapia** [comment]
My only comment is that whatever decision is made please ensure its consistent between live and backtest mode and that it is well documented somewhere.

**Comment by jshridha** [comment]
I think that consistency and backwards compatibility are key. Users have already developed algorithms assuming that last candle is a partial candle. I think that should remain the way it works and users can drop the last candle if needed. Regardless of what is decided, this should definitely be documented to prevent future confusion.

**Comment by lenak25** [comment]
@xbionicx your reasoning is correct.
@gatapia and @jshridha thanks for your inputs.
We are working on this and you are right, this should also be added to the documentation.

**Comment by jshridha** [comment]
@zackgow @aetherqi It looks like this look ahead bias still exists in 0.5.3. Have you guys observed this also?

**Comment by jshridha** [comment]
Correction - this looks to only be happening when requesting daily candles.

**Comment by gatapia** [comment]
I am trying 0.5.4 (develop branch) and requesting 2 bars is still returning 3 bars using this code:
`data.history(symbol('btc_usd'), ['close', 'volume'], bar_count=2, frequency='5T')`.

This wasn't resolved on 0.5.4?  What is the expected behaviour now?

**Comment by lenak25** [comment]
Hi @gatapia,

This issue was not yet resolved, we are working on this and will update when the issue is fixed.

Thanks,
Lena

**Comment by lenak25** [comment]
Hi, 

The issue was fixed at release 0.5.9. In backtest, as in live, in minute mode, the last candle returned is a partial one. In addition, the number of candles returned will always be equal to the number requested. 
In release 0.5.9 we have also added an API reference documentation which also includes the description of this behavior. 

Feel free to let us know if there are still unclear issues. 

Thanks,
Lena
